panic found in kvclient

## Bug Report

```
panic: unlocking a not locked range

goroutine 3727 [running]:
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc07bf4ae40, 0xc020d66a00, 0x5, 0x5)
	/go/pkg/mod/go.uber.org/zap@v1.16.0/zapcore/entry.go:234 +0x735
go.uber.org/zap.(*Logger).Panic(0xc081ea4180, 0x53ee17c, 0x1c, 0xc020d66a00, 0x5, 0x5)
	/go/pkg/mod/go.uber.org/zap@v1.16.0/logger.go:226 +0x96
github.com/pingcap/log.Panic(0x53ee17c, 0x1c, 0xc020d66a00, 0x5, 0x5)
	/go/pkg/mod/github.com/pingcap/log@v0.0.0-20201112100606-8f1e84a3abc8/global.go:50 +0xf1
github.com/pingcap/ticdc/pkg/regionspan.(*RegionRangeLock).UnlockRange(0xc041c8dcb0, 0xc041eac960, 0x12, 0x12, 0xc041eac980, 0x12, 0x12, 0xb8, 0x22, 0x5df75bab0980001)
	/home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/ticdc/pkg/regionspan/region_range_lock.go:321 +0x28d7
github.com/pingcap/ticdc/cdc/kv.(*eventFeedSession).onRegionFail(0xc03b4377a0, 0x5ef5ae0, 0xc03c154b00, 0xb8, 0x19, 0x22, 0xc041eac960, 0x12, 0x12, 0xc041eac980, ...)
	/home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/ticdc/cdc/kv/client.go:650 +0x504
github.com/pingcap/ticdc/cdc/kv.(*eventFeedSession).dispatchRequest(0xc03b4377a0, 0x5ef5ae0, 0xc03c154b00, 0xc041c8df80, 0xc000000001, 0x54b)
	/home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/ticdc/cdc/kv/client.go:707 +0x34b3
github.com/pingcap/ticdc/cdc/kv.(*eventFeedSession).eventFeed.func1(0xc010eacf68, 0x1)
	/home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/ticdc/cdc/kv/client.go:558 +0xa1
golang.org/x/sync/errgroup.(*Group).Go.func1(0xc041c8df80, 0xc041c8dfb0)
	/go/pkg/mod/golang.org/x/sync@v0.0.0-20201020160332-67f06af15bc9/errgroup/errgroup.go:57 +0x72
created by golang.org/x/sync/errgroup.(*Group).Go
	/go/pkg/mod/golang.org/x/sync@v0.0.0-20201020160332-67f06af15bc9/errgroup/errgroup.go:54 +0x74

```
In PR: https://github.com/pingcap/ticdc/pull/1187
CI link: https://internal.pingcap.net/idc-jenkins/blue/organizations/jenkins/cdc_ghpr_integration_test/detail/cdc_ghpr_integration_test/4557/pipeline/85

logs:
[processor_panic.zip](https://github.com/pingcap/ticdc/files/6048470/processor_panic.zip)

It is an old bug, v4.0.10 also has similar panic

```
panic: unlocking a not locked range

goroutine 846 [running]:
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc0ccc00300, 0xc1198d9a40, 0x5, 0x5)
        go.uber.org/zap@v1.16.0/zapcore/entry.go:234 +0x585
go.uber.org/zap.(*Logger).Panic(0xc11a19ac60, 0x284acf6, 0x1c, 0xc1198d9a40, 0x5, 0x5)
        go.uber.org/zap@v1.16.0/logger.go:226 +0x85
github.com/pingcap/log.Panic(0x284acf6, 0x1c, 0xc1198d9a40, 0x5, 0x5)
        github.com/pingcap/log@v0.0.0-20201112100606-8f1e84a3abc8/global.go:50 +0x10b
github.com/pingcap/ticdc/pkg/regionspan.(*RegionRangeLock).UnlockRange(0xc002f85890, 0xc1021e5820, 0x1b, 0x20, 0xc1021e5840, 0x1b, 0x20, 0x2f65a4e, 0xc697, 0x5de3246ab1c000a)
        github.com/pingcap/ticdc/pkg/regionspan/region_range_lock.go:321 +0x2145
github.com/pingcap/ticdc/cdc/kv.(*eventFeedSession).onRegionFail(0xc00097e400, 0x2c2d0a0, 0xc002fa45c0, 0x2f65a4e, 0x320, 0xc697, 0xc1021e5820, 0x1b, 0x20, 0xc1021e5840, ...)
        github.com/pingcap/ticdc/cdc/kv/client.go:613 +0x335
github.com/pingcap/ticdc/cdc/kv.(*eventFeedSession).dispatchRequest(0xc00097e400, 0x2c2d0a0, 0xc002fa45c0, 0xc002f85b00, 0x12, 0x12)
        github.com/pingcap/ticdc/cdc/kv/client.go:670 +0x245a
github.com/pingcap/ticdc/cdc/kv.(*eventFeedSession).eventFeed.func1(0x2c2d0a0, 0xc00273dd80)
        github.com/pingcap/ticdc/cdc/kv/client.go:521 +0x45
golang.org/x/sync/errgroup.(*Group).Go.func1(0xc002f85b00, 0xc002f85b30)
        golang.org/x/sync@v0.0.0-20201020160332-67f06af15bc9/errgroup/errgroup.go:57 +0x59
created by golang.org/x/sync/errgroup.(*Group).Go
        golang.org/x/sync@v0.0.0-20201020160332-67f06af15bc9/errgroup/errgroup.go:54 +0x66
panic: tikv reported duplicated request to the same region, which is not expected

goroutine 4736042 [running]:
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc0a38aa780, 0xc0aba8ef80, 0x1, 0x1)
        go.uber.org/zap@v1.16.0/zapcore/entry.go:234 +0x585
go.uber.org/zap.(*Logger).Panic(0xc0b52ff680, 0x28c24fe, 0x4a, 0xc0aba8ef80, 0x1, 0x1)
        go.uber.org/zap@v1.16.0/logger.go:226 +0x85
github.com/pingcap/log.Panic(0x28c24fe, 0x4a, 0xc0aba8ef80, 0x1, 0x1)
        github.com/pingcap/log@v0.0.0-20201112100606-8f1e84a3abc8/global.go:50 +0x10b
github.com/pingcap/ticdc/cdc/kv.(*eventFeedSession).handleError(0xc0211ee000, 0x2c2d0a0, 0xc029604300, 0x41bee27, 0x1b2, 0x975e, 0xc0a2c06920, 0x1b, 0x20, 0xc0a2c06960, ...)
        github.com/pingcap/ticdc/cdc/kv/client.go:1004 +0x90c
github.com/pingcap/ticdc/cdc/kv.(*eventFeedSession).eventFeed.func3(0x10, 0x0)
        github.com/pingcap/ticdc/cdc/kv/client.go:546 +0x1e5
golang.org/x/sync/errgroup.(*Group).Go.func1(0xc036d92030, 0xc0359b8d80)
        golang.org/x/sync@v0.0.0-20201020160332-67f06af15bc9/errgroup/errgroup.go:57 +0x59
created by golang.org/x/sync/errgroup.(*Group).Go
        golang.org/x/sync@v0.0.0-20201020160332-67f06af15bc9/errgroup/errgroup.go:54 +0x66
```

Need investigation

--- a/cdc/kv/client.go
+++ b/cdc/kv/client.go
@@ -1053,6 +1053,7 @@ func (s *eventFeedSession) handleError(ctx context.Context, errInfo regionErrorI
 		}
 	}
 
+	failpoint.Inject("kvClientRegionReentrantErrorDelay", nil)
 	s.scheduleRegionRequest(ctx, errInfo.singleRegionInfo)
 	return nil
 }

--- a/cdc/kv/client_test.go
+++ b/cdc/kv/client_test.go
@@ -2321,3 +2321,80 @@ func (s *etcdSuite) TestResolveLockNoCandidate(c *check.C) {
 	wg2.Wait()
 	cancel()
 }
+
+// TestFailRegionReentrant tests one region could be failover multiple times,
+// kv client must avoid duplicated `onRegionFail` call for the same region.
+// In this test
+// 1. An `unknownErr` is sent to kv client first to trigger `handleSingleRegionError` in region worker.
+// 2. We delay the kv client to re-create a new region request by 500ms via failpoint.
+// 3. Before new region request is fired, simulate kv client `stream.Recv` returns an error, the stream
+//    handler will signal region worker to exit, which will evict all active region states then.
+func (s *etcdSuite) TestFailRegionReentrant(c *check.C) {
+	defer testleak.AfterTest(c)()
+	defer s.TearDownTest(c)
+	ctx, cancel := context.WithCancel(context.Background())
+	wg := &sync.WaitGroup{}
+
+	ch1 := make(chan *cdcpb.ChangeDataEvent, 10)
+	srv1 := newMockChangeDataService(c, ch1)
+	server1, addr1 := newMockService(ctx, c, srv1, wg)
+
+	defer func() {
+		close(ch1)
+		server1.Stop()
+		wg.Wait()
+	}()
+
+	rpcClient, cluster, pdClient, err := mocktikv.NewTiKVAndPDClient("")
+	c.Assert(err, check.IsNil)
+	pdClient = &mockPDClient{Client: pdClient, versionGen: defaultVersionGen}
+	tiStore, err := tikv.NewTestTiKVStore(rpcClient, pdClient, nil, nil, 0)
+	c.Assert(err, check.IsNil)
+	kvStorage := newStorageWithCurVersionCache(tiStore, addr1)
+	defer kvStorage.Close() //nolint:errcheck
+
+	regionID := uint64(3)
+	cluster.AddStore(1, addr1)
+	cluster.Bootstrap(regionID, []uint64{1}, []uint64{4}, 4)
+
+	err = failpoint.Enable("github.com/pingcap/ticdc/cdc/kv/kvClientRegionReentrantError", "1*return(\"ok\")->1*return(\"error\")")
+	c.Assert(err, check.IsNil)
+	err = failpoint.Enable("github.com/pingcap/ticdc/cdc/kv/kvClientRegionReentrantErrorDelay", "sleep(500)")
+	c.Assert(err, check.IsNil)
+	defer func() {
+		_ = failpoint.Disable("github.com/pingcap/ticdc/cdc/kv/kvClientStreamRecvError")
+		_ = failpoint.Disable("github.com/pingcap/ticdc/cdc/kv/kvClientStreamRecvErrorDelay")
+	}()
+	baseAllocatedID := currentRequestID()
+	lockresolver := txnutil.NewLockerResolver(kvStorage.(tikv.Storage))
+	isPullInit := &mockPullerInit{}
+	cdcClient := NewCDCClient(ctx, pdClient, kvStorage.(tikv.Storage), &security.Credential{})
+	eventCh := make(chan *model.RegionFeedEvent, 10)
+	wg.Add(1)
+	go func() {
+		err := cdcClient.EventFeed(ctx, regionspan.ComparableSpan{Start: []byte("a"), End: []byte("b")}, 100, false, lockresolver, isPullInit, eventCh)
+		c.Assert(errors.Cause(err), check.Equals, context.Canceled)
+		cdcClient.Close() //nolint:errcheck
+		wg.Done()
+	}()
+
+	// wait request id allocated with: new session, new request
+	waitRequestID(c, baseAllocatedID+1)
+	unknownErr := &cdcpb.ChangeDataEvent{Events: []*cdcpb.Event{
+		{
+			RegionId:  3,
+			RequestId: currentRequestID(),
+			Event: &cdcpb.Event_Error{
+				Error: &cdcpb.Error{},
+			},
+		},
+	}}
+	ch1 <- unknownErr
+	// use a fake event to trigger one more stream.Recv
+	initialized := mockInitializedEvent(regionID, currentRequestID())
+	ch1 <- initialized
+	// since re-establish new region request is delayed by `kvClientRegionReentrantErrorDelay`
+	// there will be reentrant region failover, the kv client should not panic.
+	time.Sleep(time.Second)
+	cancel()
+}

--- a/cdc/kv/client_v2.go
+++ b/cdc/kv/client_v2.go
@@ -205,6 +205,11 @@ func (s *eventFeedSession) receiveFromStreamV2(
 	for {
 		cevent, err := stream.Recv()
 
+		failpoint.Inject("kvClientRegionReentrantError", func(op failpoint.Value) {
+			if op.(string) == "error" {
+				worker.inputCh <- nil
+			}
+		})
 		failpoint.Inject("kvClientStreamRecvError", func() {
 			err = errors.New("injected stream recv error")
 		})

--- a/cdc/kv/region_worker.go
+++ b/cdc/kv/region_worker.go
@@ -96,6 +96,10 @@ func (w *regionWorker) handleSingleRegionError(ctx context.Context, err error, s
 		zap.Stringer("span", state.sri.span),
 		zap.Uint64("checkpoint", state.sri.ts),
 		zap.String("error", err.Error()))
+	// if state is already marked stopped, it must have been or would be processed by `onRegionFail`
+	if state.isStopped() {
+		return nil
+	}
 	// We need to ensure when the error is handled, `isStopped` must be set. So set it before sending the error.
 	state.markStopped()
 	failpoint.Inject("kvClientSingleFeedProcessDelay", nil)
@@ -443,9 +447,15 @@ func (w *regionWorker) evictAllRegions(ctx context.Context) error {
 	w.statesLock.Lock()
 	defer w.statesLock.Unlock()
 	for _, state := range w.regionStates {
-		state.lock.RLock()
+		state.lock.Lock()
+		// if state is marked as stopped, it must have been or would be processed by `onRegionFail`
+		if state.isStopped() {
+			state.lock.Unlock()
+			continue
+		}
+		state.markStopped()
 		singleRegionInfo := state.sri.partialClone()
-		state.lock.RUnlock()
+		state.lock.Unlock()
 		err := w.session.onRegionFail(ctx, regionErrorInfo{
 			singleRegionInfo: singleRegionInfo,
 			err: &rpcCtxUnavailableErr{

